File: update_frontmatter.py
import frontmatter
import os
from datetime import datetime, date

def parse_date(date_str):
    if isinstance(date_str, datetime):
        return date_str
    elif isinstance(date_str, date):
        return datetime(date_str.year, date_str.month, date_str.day)

    formats = ['%Y-%m-%d', '%Y-%m-%dT%H:%M:%S.%fZ', '%Y-%m-%dT%H:%M:%SZ', '%Y-%m-%dT%H:%M', '%Y-%m-%d %H:%M']
    for fmt in formats:
        try:
            return datetime.strptime(date_str, fmt)
        except ValueError:
            pass
    raise ValueError(f"Date format not recognized: {date_str}")

def update_tags(post, today):
    try:
        creation_date = parse_date(post['date'])
    except ValueError as e:
        print(e)
        return False

    diff = today - creation_date
    new_tag = None

    if diff.days >= 42:
        new_tag = 'final'
    elif diff.days >= 28:
        new_tag = 'accepted'
    elif diff.days >= 14:
        new_tag = 'last call'

    if new_tag:
        tags = post.get('tags', [])
        if new_tag not in tags:
            tags = [tag for tag in tags if tag not in ['draft', 'last call', 'accepted', 'final']]
            tags.append(new_tag)
            post['tags'] = tags
            return True
    return False

def process_markdown_files():
    for subdir, dirs, files in os.walk('cip'):
        for file in files:
            if file.endswith('.md') or file.endswith('.mdx'):
                filepath = os.path.join(subdir, file)
                with open(filepath, 'r', encoding='utf-8') as f:
                    post = frontmatter.load(f)
                    if update_tags(post, datetime.now()):
                        with open(filepath, 'w', encoding='utf-8') as f:
                            frontmatter.dump(post, f)

if __name__ == "__main__":
    process_markdown_files()

File: update_frontmatter.mjs
import { glob } from 'glob';
import fs from 'fs';
import matter from 'gray-matter';

function parseDate(dateStr) {
	if (!dateStr) {
		console.warn('No date found in frontmatter');
		return null;
	}
	const date = new Date(dateStr);
	if (isNaN(date.getTime())) {
		console.warn(`Invalid date format: ${dateStr}`);
		return null;
	}
	return date;
}

function updateStatus(fileContent, today) {
	try {
		const post = matter(fileContent);

		// Check if frontmatter exists and has required fields
		if (!post.data || !post.data.date) {
			console.warn('No frontmatter or date found');
			return null;
		}

		const creationDate = parseDate(post.data.date);
		if (!creationDate) {
			return null;
		}

		const diff = (today - creationDate) / (1000 * 60 * 60 * 24); // Difference in days
		let newStatus = 'draft';

		if (diff >= 42) {
			newStatus = 'final';
		} else if (diff >= 28) {
			newStatus = 'accepted';
		} else if (diff >= 14) {
			newStatus = 'last call';
		}

		if (post.data.status !== newStatus) {
			post.data.status = newStatus;
			return matter.stringify(post.content, post.data);
		}
		return null;
	} catch (error) {
		console.error(`Error processing file: ${error.message}`);
		return null;
	}
}

async function processMarkdownFiles() {
	try {
		const files = await glob('cip/**/*.md');
		console.log(`Found ${files.length} markdown files to process`);

		for (const file of files) {
			try {
				const content = fs.readFileSync(file, 'utf8');
				const today = new Date();
				const updatedContent = updateStatus(content, today);

				if (updatedContent) {
					fs.writeFileSync(file, updatedContent, 'utf8');
					console.log(`Updated status in: ${file}`);
				}
			} catch (error) {
				console.error(`Error processing file ${file}: ${error.message}`);
			}
		}
	} catch (error) {
		console.error(`Error finding markdown files: ${error.message}`);
		process.exit(1);
	}
}

processMarkdownFiles().catch(error => {
	console.error(`Script failed: ${error.message}`);
	process.exit(1);
});
